### DISCUSSION TITLE
Custom error handling

### DISCUSSION THREAD

**Discussion by ChaplyginWezom** [nuwave/lighthouse#2521] [Q&A]
Hi! I've tried to write my own custom error handler, so I can override message given from Laravel's `ModelNotFoundException`.
But, in my case exception is being thrown inside of `@canFind` directive, then catched and completely replaced with generic Webonyx `Error`
```php
            $modelOrModels = $this->directiveArgValue('findOrFail', true)
                    ? $enhancedBuilder->findOrFail($findValue)
                    : $enhancedBuilder->find($findValue);
            } catch (ModelNotFoundException $modelNotFoundException) {
                throw new Error($modelNotFoundException->getMessage());
            }
```
Is it intended behaviour due to some security concerns or was it just overlooked feature? Because I saw that in some earlier version of Lighthouse this exception was never catched for `@can` directive

**Comment by spawnia** [comment]
See https://github.com/nuwave/lighthouse/pull/1225.

**Reply by ChaplyginWezom** [reply]
Okay, so basically if I want to give user-friendly translated error message like "Record(s) not found", I should use `findOrFail` in concrete resolver instead of `@canFind` directive
This slightly limits usage of such directive, but okay, at least i can extend it if I need to implement this behaviour globally
Thanks!

**Reply by spawnia** [reply]
Wrapping in `Error` actually ensures that the error message is shown to the end user.

**Reply by ChaplyginWezom** [reply]
Yes, but as you can see in the example above - that error doesn't mention previous `ModelNotFoundException` thus making custom handler inapplicable for such case (if not to resort for exception identification using regex on message...)

Actually, you reminded me why I wanted to start this discussion in the first place: `GraphQL\Error` doen't receive previous exception when being thrown in **can*** directives, leading to problems mentioned above
Please, correct if I'm wrong, but I thought that in such case exception should have contain previous one

**Reply by spawnia** [reply]
Adding the previous exception makes sense. I am open to a pull request that changes this.

**Reply by ChaplyginWezom** [reply]
Unfortunately, I'm struggling to pass respective tests, as linking previous exception makes it not client-safe
```php
$this->isClientSafe = $previous instanceof ClientAware
    ? $previous->isClientSafe()
    : $previous === null;
```
From my understanding the only way to circumvent this is to provide respective error handler for `ModelNotFoundException` from the start. But implementing such feature will build stronger dependency on Laravel framework. So this is a dead end on my side...

**Reply by spawnia** [reply]
We don't have to throw `GraphQL\Error`, we can introduce a custom `ClientSafeModelNotFoundException`.

**Reply by ChaplyginWezom** [reply]
You're right. And then, with passed previous exception, developer could fine-tune desired message format in own handler.
The last question I have: should we change default message of a new exception so it can be a **truly** client-safe exception?)

I mean default model exception maybe do not expose sensitive information, but renders class name with namespace, which is probably not very nice for production environment. Or it would be an unnecessary side-effect in taking such precautions?

**Reply by spawnia** [reply]
I am a bit wary if massaging the Laravel error message or customizing it is useful or necessary. Let's do the quick fix first, add a test that covers it, then we can maybe iterate on better messages.
